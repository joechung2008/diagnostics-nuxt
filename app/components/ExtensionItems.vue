<template>
  <div class="extension-root">
    <UNavigationMenu orientation="vertical" :items="navItems" />
  </div>
</template>

<script setup lang="ts">
import type { NavigationMenuItem } from "@nuxt/ui";
import type { ExtensionsProps, KeyedNavLink } from "~/types";
import { byKey, isExtensionInfo, toNavLink } from "~/utils/extensions";

const props = defineProps<ExtensionsProps>();

const links = computed(() => {
  return Object.values(props.extensions)
    .filter(isExtensionInfo)
    .map(toNavLink)
    .sort(byKey);
});

const navItems = computed<NavigationMenuItem[]>(() => {
  return links.value.map((link: KeyedNavLink) => ({
    label: link.name,
    onSelect: (e: MouseEvent) => {
      e.preventDefault();
      props.onLinkClick(e, link);
    },
  }));
});
</script>
