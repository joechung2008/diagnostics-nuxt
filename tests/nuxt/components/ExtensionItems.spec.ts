import { mountSuspended, renderSuspended } from "@nuxt/test-utils/runtime";
import { describe, expect, it, vi } from "vitest";
import ExtensionItems from "../../../app/components/ExtensionItems.vue";
import type {
  ExtensionError,
  ExtensionInfo,
  ExtensionsProps,
} from "../../../app/types";

describe("app/components/ExtensionItems.vue", () => {
  it("should mount", async () => {
    const component = await mountSuspended(ExtensionItems);
    expect(component).toBeDefined();
  });

  it("should render extension items", async () => {
    const extension1: ExtensionInfo = {
      extensionName: "Database Extension",
      config: {
        host: "localhost",
        port: "5432",
      },
    };

    const extension2: ExtensionError = {
      lastError: {
        errorMessage: "Connection timeout",
        time: "2025-01-01T10:00:00Z",
      },
    };

    const props: ExtensionsProps = {
      extensions: {
        db: extension1,
        auth: extension2,
      },
      onLinkClick: () => {},
    };

    const { html } = await renderSuspended(ExtensionItems, {
      props,
    });

    expect(html()).toMatchSnapshot();
  });

  it("should call onLinkClick when an extension link is clicked", async () => {
    const extension1: ExtensionInfo = {
      extensionName: "Database Extension",
      config: {
        host: "localhost",
        port: "5432",
      },
    };

    const mockOnLinkClick = vi.fn();
    const props: ExtensionsProps = {
      extensions: {
        db: extension1,
      },
      onLinkClick: mockOnLinkClick,
    };

    const component = await mountSuspended(ExtensionItems, {
      props,
    });

    // Find the navigation menu item containing the extension link
    const navMenuItem = component.find(
      '.extension-root nav [data-testid="nav-menu-item"], .extension-root [role="menuitem"]'
    );

    if (navMenuItem.exists()) {
      // Simulate clicking the menu item
      await navMenuItem.trigger("click");

      // Verify that onLinkClick was called with the correct arguments
      expect(mockOnLinkClick).toHaveBeenCalledTimes(1);

      // Check the arguments: event (MouseEvent) and link (KeyedNavLink)
      const [eventArg, linkArg] = mockOnLinkClick.mock.calls[0];
      expect(eventArg).toBeInstanceOf(MouseEvent);
      expect(linkArg).toEqual({
        key: "Database Extension",
        name: "Database Extension",
      });
    } else {
      // Fallback: Check for any clickable element if specific selector fails
      const clickableElements = component.findAll(
        'button, a, [role="button"], [tabindex="0"]'
      );
      if (clickableElements.length > 0) {
        await clickableElements[0].trigger("click");
        expect(mockOnLinkClick).toHaveBeenCalledTimes(1);
      }
    }
  });
});
