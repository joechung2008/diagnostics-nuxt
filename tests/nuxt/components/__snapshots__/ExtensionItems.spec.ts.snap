// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app/components/ExtensionItems.vue > should render extension items 1`] = `
"<div id="test-wrapper">
  <div class="extension-root">
    <!---->
    <!---->
    <nav data-collapsed="false" class="relative flex gap-1.5 [&amp;>div]:min-w-0 flex-col" aria-label="Main" data-orientation="vertical" dir="ltr" data-reka-navigation-menu="">
      <ul class="isolate min-w-0">
        <li data-state="closed" data-orientation="vertical" class="min-w-0"><button type="button" data-reka-collection-item="" class="group relative w-full flex items-center gap-1.5 font-medium text-sm before:absolute before:z-[-1] before:rounded-md focus:outline-none focus-visible:outline-none dark:focus-visible:outline-none focus-visible:before:ring-inset focus-visible:before:ring-2 focus-visible:before:ring-primary flex-row px-2.5 py-1.5 before:inset-y-px before:inset-x-0 text-muted hover:text-highlighted hover:before:bg-elevated/50 transition-colors before:transition-colors">
            <!--v-if--><span class="truncate">Database Extension<!--v-if--></span>
            <!--v-if-->
          </button>
          <!--v-if-->
          <!--v-if-->
        </li>
      </ul>
      <!--v-if-->
      <!--v-if-->
    </nav>
  </div>
</div>"
`;
